### PR TITLE
Feature interface CRTP vervangen

### PR DESCRIPTION
**Pull request by Slimmerik**
CRTP vervangen 
added base classes for uarm and 4dof diy arm

### CODE REVIEW

File: Makefile.due [hunk 1]
```diff
@@ -10,7 +10,7 @@
 
 # settings for Arduino Due projects
 TARGET            ?= arduino_due
-SERIAL_PORT       ?= COM2
+SERIAL_PORT       ?= COM3
```
Comment (Timmotus) [inline Makefile.due]: this should probably not be changed

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+set(CMAKE_CXX_STANDARD 17)
+add_executable(R2D2 ./code/main.cpp code/headers/robot_arm_interface_c.hpp code/headers/uArm_swift_pro_c.hpp code/src/uArm_swift_pro_c.cpp code/headers/dof4_diy_c.hpp code/src/dof4_diy_c.cpp)
+add_executable(R2D2-Test ./test/main.cpp)
+add_definitions(-D__SAM3X8E__)
+include_directories(./code/headers
+        ../../libraries/hwlib/library
+        ../../programs/bmptk/targets/cortex/atmel
+        )
+
+target_include_directories(R2D2-Test PUBLIC ../../libraries/Catch2/single_include/catch2)
```
Comment (Timmotus) [inline CMakeLists.txt]: why is this file committed?

File: README.md [hunk 3]
```diff
@@ -1 +1,49 @@
...
+# Tests
+
+## Video demonstration
+
+## Unit tests
+
+# Development log
+
+# Research Summary
+link to research page: https://github.com/R2D2-2019/R2D2-2019/wiki/Robot-Arm
```
Comment (Timmotus) [inline README.md]: this file needs to be a separate pull request as its not part of the interface

File: code/headers/robot_arm_interface_c.hpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+#include <vector3_c.hpp>
+#include <hwlib.hpp>
+
+namespace r2d2::robot_arm{
+     class robot_arm_interface{
+
+     public:
+         // pure virtual function providing interface framework.
+         virtual void move_head_to_coordinate(const vector3_c< uint16_t > &v) = 0;
```
Comment (Timmotus) [inline code/headers/robot_arm_interface_c.hpp]: use a more descriptive variable name and use type vector3i_c instead of vector3_c<uint16_t>

File: code/headers/robot_arm_interface_c.hpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+         virtual void rotate_head(const uint16_t &r) =0;
```
Comment (Timmotus) [inline code/headers/robot_arm_interface_c.hpp]: use a more descriptive variable name

File: code/headers/robot_arm_interface_c.hpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+     protected:
```
Comment (Timmotus) [inline code/headers/robot_arm_interface_c.hpp]: protected should be removed if emtpy

File: code/headers/uArm_swift_pro_c.hpp [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace r2d2::robot_arm{
+    class uArm_swift_pro_c : public robot_arm_interface{
+    public:
+
+        void move_head_to_coordinate(const vector3_c< uint16_t > &v);
+
+        void rotate_head(const uint16_t &r);
+
+    private:
```
Comment (Timmotus) [inline code/headers/uArm_swift_pro_c.hpp]: private should be removed if empty

File: code/headers/dof4_diy_c.hpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace r2d2::robot_arm{
+    class dof4_diy : public robot_arm_interface{
+    public:
+
+        void move_head_to_coordinate(const vector3_c< uint16_t > &v);
+
+        void rotate_head(const uint16_t &r);
+
+    private:
```
Comment (Timmotus) [inline code/headers/dof4_diy_c.hpp]: private should be removed if empty

**Review by Timmotus — CHANGES_REQUESTED** [review]
Doxygen needs to be added.
Own files should be included using quotes instead of angled brackets.
"Created by" should be removed

File: code/headers/dof4_diy_c.hpp [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+#include <robot_arm_interface_c.hpp>
+#include <vector3_c.hpp>
+
+
+
+namespace r2d2::robot_arm{
+    class dof4_diy : public robot_arm_interface{
+    public:
+
+        void move_head_to_coordinate(const vector3_c< uint16_t > &v);
```
Comment (Kaaserne) [inline code/headers/dof4_diy_c.hpp]: Please add `override` if method overrides function from interface.

File: code/headers/robot_arm_interface_c.hpp (hunk 4 above)
Comment (Kaaserne) [reply code/headers/robot_arm_interface_c.hpp]: Also no spaces between angle brackets, use `<uint16_t>` instead of `< uint16_t >`. Use `vector3i_c` for integer values. Maybe negative values can be passed with other arms.

File: code/headers/dof4_diy_c.hpp [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+        void rotate_head(const uint16_t &r);
```
Comment (Kaaserne) [inline code/headers/dof4_diy_c.hpp]: Please add override if method `overrides` function from interface. Also, I believe that `uint16_t` should be `int16_t` since a negative value can be passed.

**Review by Kaaserne — CHANGES_REQUESTED** [review]
Please run .clang format over all of the relatable files.

File: code/headers/uArm_swift_pro_c.hpp [hunk 11]
```diff
@@ -1,23 +1,29 @@
...
-
-        void rotate_head(const uint16_t &r);
-
-    private:
+        /**
+         * This function moves the uArm swift pro head to a certain 3d location.
+         *
+         * @param v
+         */
+        void move_head_to_coordinate(const vector3i_c &vector) override;
```
Comment (Timmotus) [inline code/headers/uArm_swift_pro_c.hpp]: vector variable name should be more along the lines of 'coordinate'

File: code/headers/uArm_swift_pro_c.hpp [hunk 12]
```diff
@@ -1,23 +1,29 @@
-//
-// Created by Slimmerik on 13/04/2019.
-//
+/**
+ * Class uArm_swift_pro provides a implementation for uArm swift pro.
+ */
+
```
Comment (Timmotus) [inline code/headers/uArm_swift_pro_c.hpp]: this part of doxygen should be just above the class, not at the top of the document

**Review by Timmotus — CHANGES_REQUESTED** [review]

**Review by Timmotus — APPROVED** [review]

**Review by Kaaserne — APPROVED** [review]
Looks good
